mscore/plugin/pluginUpdater.cpp

```diff
@@ -0,0 +1,840 @@
+                  if (fi.isFile && QFileInfo(fi.filePath).suffix() == "qml")
+                        has_qml = true;
+                  }
+            if (!has_qml) {
+                  // invalid plugin
+                  if (result)
+                        *result = INSTALL_FAILED_INVALID;
+                  QFile::remove(download_pkg);
+                  return false;
+                  }
+            if (dirs.size() == 1)
+                  if (allFiles.size() > 1) // the element in dirs must be a dir then
+                        has_no_dir = false;
+            int stripped_len = 0;
+            if (!has_no_dir) {
+                  stripped_len = allFiles.first().filePath.length() + 1;
+                  // the directory in the archive is not to be added
+                  allFiles.pop_front();
+                  }
+            auto filePathStripper = [&](QString& filePath) {return filePath.right(filePath.size() - stripped_len); };
+
+            // extract and copy
+            for (MQZipReader::FileInfo fi : allFiles) {
+                  if (fi.filePath.startsWith("__MACOSX"))
+                        continue;
+                  if (fi.isDir)
+                        QDir().mkdir(destination_prefix + "/" + filePathStripper(fi.filePath));
+                  else if (fi.isFile) {
+                        if (QFileInfo(fi.filePath).suffix().toLower() == "qml")
+                              desc.qml_paths.push_back(destination_prefix + "/" + filePathStripper(fi.filePath));
+                        QFile new_f(destination_prefix + "/" + filePathStripper(fi.filePath));
+                        if (!new_f.open(QIODevice::WriteOnly)) {
+                              qDebug("Cannot write the file.");
+                              return false;
+                              }
+                        new_f.write(zipFile.fileData(fi.filePath));
+                        new_f.setPermissions(fi.permissions);
+                        new_f.close();
+                        }
+                  }
+            zipFile.close();
+            }
+      else if (suffix == "qml") {
+            QString destination = destination_prefix + "/" + f_pkg.fileName();
+            QFileInfo check(destination);
+            if (check.exists()) {
+                  if (check.isDir())
+                        QDir(check.absoluteFilePath()).removeRecursively();
+                  if (check.isFile())
+                        QFile::remove(check.absoluteFilePath());
+                  }
+            QFile(f_pkg.absoluteFilePath()).copy(destination);
+            desc.qml_paths.push_back(destination);
+            }
+      else {
+            qDebug("Unknown plugin suffix %s.", qPrintable(suffix));
+            return false;
+            }
+      QFile::remove(download_pkg);
+      return true;
+      }
+
+void PluginWorker::downloadInstall(QPushButton* button)
+      {
+      button->setEnabled(false);
+      button->setText(tr("Analyzing…"));
```

Comment:
and here

Reply:
I was lazy and thought this state transition didn't change much(only button text is changed). But reusing the string is indeed safer(it avoids accessing `button` directly by using signals and slots, so won't crash even if the `button` becomes `nullptr`), so I'll make the change, thanks!